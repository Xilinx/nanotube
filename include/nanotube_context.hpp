/**************************************************************************\
*//*! \file nanotube_context.hpp
** \author  Neil Turton <neilt@amd.com>
**  \brief  Implementation of Nanotube contexts.
**   \date  2020-09-01
*//*
\**************************************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/

#ifndef NANOTUBE_CONTEXT_HPP
#define NANOTUBE_CONTEXT_HPP

#include "nanotube_api.h"

#include <map>
#include <unordered_map>
#include <utility>

class nanotube_thread;

/*! A Nanotube context for holding resources.
 *
 * A Nanotube context is used for passing resources to a packet kernel
 * or a thread.
 *
 * This class needs to be merged with nanotube_context, but that
 * struct is currently visible to the kernel written in C.
 */
class nanotube_context
{
public:
  nanotube_context();
  ~nanotube_context();

  /*! Bind a thread to a context. */
  void bind_thread(nanotube_thread *thread);

  /*! Unbind a thread from a context. */
  void unbind_thread(nanotube_thread *thread);

  /*! Determine whether the associated thread is the current thread. */
  bool is_current();

  /*! Determine whether the associated thread is stopped. */
  bool is_stopped();

  /*! Add a channel to the context.
   *
   * \param channel_id The ID which will be used to access the channel.
   * \param channel    The channel to add.
   * \param flags      The types of access which need to be supported.
   */
  void add_channel(nanotube_channel_id_t channel_id,
                   nanotube_channel* channel,
                   nanotube_channel_flags_t flags);

  /*! Retrieve a channel from the context.
   *
   * \param channel_id  The ID which was passed to add_channel.
   * \param flags       The type of access which is requested.
   *
   * \return            A reference to the channel.
   *
   * This method raised an error on failure.
   */
  nanotube_channel &find_channel(nanotube_channel_id_t channel_id,
                                 nanotube_channel_flags_t flags);

  /*!
   * Add a map to the context.
   *
   * \param id    Numerical ID of the map to be added.
   * \param map   Pointer to the map that should be added.
   */
  void add_map(nanotube_map_id_t id, nanotube_map_t* map);

  /*!
   * Get the specified map; returns nullptr in case the id was not found.
   * \param id    The ID of the map that should be returned.
   * \return      Pointer to the map.
   */
  nanotube_map_t* get_map(nanotube_map_id_t id);

  /*! Check that this context can be used from the current thread. 
   *
   * This method raises a fatal error if the context cannot be used
   * from the current thread.
   */
  void check_thread();

  /*! Wake up the thread associated with this context. */
  void wake();

  nanotube_thread* get_thread() {return m_thread;}

private:
  typedef std::pair<nanotube_channel_id_t, nanotube_channel_flags_t>
    port_info_t;

  // The thread associated with the context, if any.
  nanotube_thread *m_thread;

  std::map<port_info_t, nanotube_channel_t *>            m_channels;
  std::unordered_map<nanotube_map_id_t, nanotube_map_t*> m_maps;
};

#endif // NANOTUBE_CONTEXT_HPP
