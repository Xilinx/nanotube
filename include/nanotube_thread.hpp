/**************************************************************************\
*//*! \file nanotube_thread.hpp
** \author  Neil Turton <neilt@amd.com>
**  \brief  Implementation of Nanotube threads.
**   \date  2020-09-03
*//*
\**************************************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/

#ifndef NANOTUBE_THREAD_HPP
#define NANOTUBE_THREAD_HPP

#include "nanotube_api.h"

#include <atomic>
#include <memory>
#include <mutex>
#include <vector>

#include <boost/date_time/posix_time/posix_time_types.hpp>

extern "C" {
#include <pthread.h>
}

///////////////////////////////////////////////////////////////////////////

/*! A guard class for locking stderr.
**
** This class can be used to lock stderr.  Create an instance of the
** class to hold the lock for the lifetime of the instance.  A static
** member can be used to get direct access to the mutex.
*/

class nanotube_stderr_guard
{
public:
  // Create the guard instance.
  nanotube_stderr_guard(): m_guard(s_stderr_mutex) {}

  // Get access to the mutex.
  static std::mutex &stderr_mutex() { return s_stderr_mutex; }

private:
  // The stderr mutex.
  static std::mutex s_stderr_mutex;

  // 
  std::lock_guard<std::mutex> m_guard;
};

///////////////////////////////////////////////////////////////////////////

class nanotube_thread;

/*! A class for waiting until threads are idle.
**
** To wait until all of a collection of threads are idle, create an
** object of this class, call the monitor_thread method for each
** thread and then poll is_idle() to determine whether all the threads
** are idle, calling nanotube_thread_wait() after is_idle() returns
** false to wait until the monitored threads become idle.
*/
class nanotube_thread_idle_waiter
{
public:
  /*! Construct an Nanotube thread idle waiter. */
  nanotube_thread_idle_waiter();

  /*! Destroy an Nanotube thread idle waiter. */
  ~nanotube_thread_idle_waiter();

  /*! Add a thread to be monitored. */
  void monitor_thread(nanotube_thread &thread);

  /*! Check whether all the monitored threads are idle. */
  bool is_idle();

private:
  friend class nanotube_thread;

  // Increment the busy count.  This is called when a monitored thread
  // is woken.
  void inc_busy_count();

  // Decrement the busy count.  This is called when a monitored thread
  // goes to sleep.
  void dec_busy_count();

  // The number of threads which are not sleeping.
  std::atomic<size_t> m_busy_count;

  // The thread which should be woken when all monitored threads are
  // idle.
  nanotube_thread *m_waiter;

  // The threads which are being monitored.
  std::vector<nanotube_thread *> m_threads;
};

class nanotube_thread
{
public:
  // Types related to clocks.
  typedef boost::posix_time::ptime time_point_t;
  typedef boost::posix_time::time_duration duration_t;

  /*! Contruct the main thread.
  //
  // The main thread is bound to the pthread which called this
  // constructor.
  //
  // \param context The context which holds resources for the thread.
  */
  nanotube_thread(nanotube_context_t *context);

  /*! Contruct a user thread.
  //
  // A user thread will execute a function which is supplied by the
  // caller, passing it a copy of the block of data which is provided
  // by the caller.
  //
  // \param context    The context which holds resources for the thread.
  // \param name       The name of the thread.
  // \param func       The function to execute in the thread.
  // \param data       The start of the data to copy.
  // \param data_size  The number of bytes of data to copy.
  */
  nanotube_thread(nanotube_context_t *context,
                  const char *name,
                  nanotube_thread_func_t *func,
                  const void *data,
                  size_t data_size);

  /*! Destroy the thread. */
  ~nanotube_thread();

  /*! Destroy a thread and release the memory. */
  static void destory(std::unique_ptr<nanotube_thread> &&thread);

  /*! Determine whether the thread is the current thread. */
  bool is_current();

  /*! Determine whether the thread is stopped. */
  bool is_stopped();

  /*! Get the current time.
  //
  // \returns the current time.
  */
  static time_point_t get_current_time();

  /*! Initialize a timer with the specified duration.
  //
  // This call should be called to initialise the timer before calling
  // check_timer to find out whether the timer has expired.
  //
  // \param timer     The timer to initialise.
  // \param duration  The duration of the timer.
  */
  static void init_timer(time_point_t &timer,
                         const duration_t &duration);

  /*! Check whether the specified time has been reached.
  //
  // \param timer     The timer to examine.
  //
  // \returns a flag indicating whether the timer has expired.
  */
  static bool check_timer(const time_point_t &timer);

  /*! Create the underlying pthread. */
  void start();

  /*! Stop a thread and return when complete.
  // 
  // This call causes the thread to exit when convenient and waits
  // until it has done so.  It must be called from a different thread.
  */
  void stop();

  /*! Make sure this thread object is the calling thread.
  //
  // This call checks whether this thread is the called thread.  If it
  // is not, it reports a fatal error.  It also checks whether a stop
  // request is pending and causes the thread to stop if there is.
  */
  void check_current();

  /*! Causes the thread to sleep until wake() is called.
  //
  // This method causes the current thread to suspend execution and
  // resume when the wake() method is called.  If the wake() method
  // has already been called since the previous call to sleep(), this
  // method returns without sleeping.  This method must be called from
  // the associated pthread.
  */
  void sleep();

  /*! Wake a thread from its current or next call to sleep().
  //
  // This method wakes up the thread.  If the thread is currently
  // sleeping in a call to sleep() then it is woken.  If it is not
  // sleeping then the next call to sleep() will return without
  // sleeping.  Usually called from a different thread, but .
  */
  void wake();

  /*! Send a signal to the thread if it has been started.
  //
  // \param sig The signal to send.
  */
  void kill(int sig);

  std::string get_name() { return m_name; }
private:
  friend class nanotube_thread_idle_waiter;

  /*! Create the mutex and condition variable. */
  void init();

  /*! Initialize m_current_time. */
  void init_current_time();

  /*! Call the underlying Nanotube thread function. */
  static void *enter_thread(void *thread);

  /*! Call pthread_exit if an exit request is pending.
  //
  // Called from the associated pthread.
  */
  void check_stop();

  /*! Set the idle waiter for the thread.
  //
  // This method is called from the monitor_thread method of an idle
  // waiter to start monitoring the thread.  There can only be one
  // idle waiter associated with a thread at any one time.
  */
  void set_idle_waiter(nanotube_thread_idle_waiter &waiter);

  /*! Unset the idle waiter for the thread.
  //
  // This method is called from the destructor of an idle waiter to
  // stop monitoring the thread.
  */
  void unset_idle_waiter(nanotube_thread_idle_waiter &waiter);

  typedef enum thread_state: uint32_t {
    THREAD_STATE_INIT,
    THREAD_STATE_RUNNING,
    THREAD_STATE_STOP_REQ,
    THREAD_STATE_STOPPED,
  } thread_state_t;

  // Wake state transitions are as follows:
  //   Running -> Sleeping: By the thread itself, with the mutex held.
  //   Running -> Wake: By another thread.
  //   Sleeping -> Running: By the thread itself, with the mutex held.
  //   Sleeping -> Wake: By another thread, with the mutex held, must wake.
  //   Wake -> Running: By the thread itself.
  //   Wake -> Sleeping: Invalid.
  typedef enum wake_state: uint32_t {
    WAKE_STATE_RUNNING,
    WAKE_STATE_SLEEPING,
    WAKE_STATE_WAKE,
  } wake_state_t;

  // The context which holds resources for the thread.
  nanotube_context_t *m_context;

  // A flag indicating that this is the main thread.  The main thread
  // inherits an already created pthread and does not have a thread
  // function.
  bool m_is_main_thread;

  // The name of the thread.
  std::string m_name;

  // The function to enter to run the thread.
  nanotube_thread_func_t *m_func;

  // Arbitrary data which is passed to the thread function.
  std::vector<char> m_user_data;

  // The state used for starting and stopping the thread.
  std::atomic<thread_state_t> m_thread_state;

  // The state used for sleeping and for waking the thread.
  std::atomic<wake_state_t> m_wake_state;

  // A mutex which protects m_wake_state.
  pthread_mutex_t m_mutex;

  // A condition variable to wake the thread.
  pthread_cond_t m_cond;

  // The pthread ID of the thread.
  pthread_t m_thread_id;

  // Whether m_current_time is valid.
  bool m_current_time_valid;

  // The current time.
  time_point_t m_current_time;

  // Whether m_timeout is valid.
  bool m_wake_time_valid;

  // The earliest time to wake the thread.
  time_point_t m_wake_time;

  // The idle waiter for the thread.  The thread will update the idle
  // waiter whenever it transitions into or out of the SLEEPING state.
  // This member is protected by m_mutex.
  nanotube_thread_idle_waiter *m_idle_waiter;
};

///////////////////////////////////////////////////////////////////////////

#endif // NANOTUBE_THREAD_HPP
