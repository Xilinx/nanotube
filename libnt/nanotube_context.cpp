/**************************************************************************\
*//*! \file nanotube_context.cpp
** \author  Neil Turton <neilt@amd.com>
**  \brief  Implementation of Nanotube contexts.
**   \date  2020-09-01
*//*
\**************************************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/

#include "nanotube_context.hpp"

#include "nanotube_api.h"
#include "nanotube_channel.hpp"
#include "nanotube_thread.hpp"
#include "processing_system.hpp"

#include <cassert>

///////////////////////////////////////////////////////////////////////////

nanotube_context::nanotube_context():
  m_thread(nullptr)
{
}

nanotube_context::~nanotube_context()
{
  assert(m_thread == nullptr);
}

void nanotube_context::bind_thread(nanotube_thread *thread)
{
  assert(m_thread == nullptr);
  m_thread = thread;
}

void nanotube_context::unbind_thread(nanotube_thread *thread)
{
  assert(m_thread == thread);
  m_thread = nullptr;
}

bool nanotube_context::is_current()
{
  if (m_thread == nullptr)
    return false;
  return m_thread->is_current();
}

bool nanotube_context::is_stopped()
{
  if (m_thread == nullptr)
    return true;
  return m_thread->is_stopped();
}

void nanotube_context::add_channel(nanotube_channel_id_t channel_id,
                                        nanotube_channel_t* channel,
                                        nanotube_channel_flags_t flags)
{
  if (flags & NANOTUBE_CHANNEL_READ) {
    auto res = m_channels.emplace(
      std::make_pair(channel_id, NANOTUBE_CHANNEL_READ), channel);
    if (!res.second)
      throw std::runtime_error("Channel is already registered for"
                               " reading.");
    channel->set_reader(this);
  }

  if (flags & NANOTUBE_CHANNEL_WRITE) {
    auto res = m_channels.emplace(
      std::make_pair(channel_id, NANOTUBE_CHANNEL_WRITE), channel);
    if (!res.second)
      throw std::runtime_error("Channel is already registered for"
                               " writing.");
    channel->set_writer(this);
  }
}

nanotube_channel_t &
nanotube_context::find_channel(nanotube_channel_id_t channel_id,
                                    nanotube_channel_flags_t flags)
{
  auto it = m_channels.find(std::make_pair(channel_id, flags));
  assert (it != m_channels.end());
  return *(it->second);
}

void
nanotube_context::add_map(nanotube_map_id_t id, nanotube_map_t* map) {
  m_maps[id] = map;
}

nanotube_map_t*
nanotube_context::get_map(nanotube_map_id_t id) {
  auto it = m_maps.find(id);
  if( it == m_maps.end() )
    return nullptr;
  return it->second;
}

void nanotube_context::check_thread()
{
  assert(m_thread != nullptr);
  m_thread->check_current();
}

void nanotube_context::wake()
{
  assert(m_thread != nullptr);
  m_thread->wake();
}

///////////////////////////////////////////////////////////////////////////

nanotube_context_t *nanotube_context_create()
{
  auto* ctx = new nanotube_context();
  std::unique_ptr<nanotube_context> context_ptr(ctx);

  processing_system &ps = processing_system::get_current();
  ps.add_context(std::move(context_ptr));
  return ctx;
}

void nanotube_context_add_channel(nanotube_context_t* context,
                                  nanotube_channel_id_t channel_id,
                                  nanotube_channel_t* channel,
                                  nanotube_channel_flags_t flags)
{
  context->add_channel(channel_id, channel, flags);
}

void nanotube_context_add_map(nanotube_context_t* context,
                              nanotube_map_t*     map) {
  context->add_map(map->id, map);
}

///////////////////////////////////////////////////////////////////////////
