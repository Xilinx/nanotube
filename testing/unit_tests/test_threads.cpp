/*******************************************************/
/*! \file test_threads.cpp
** \author Neil Turton <neilt@amd.com>
**  \brief Unit tests for Nanotube threads implementations.
**   \date 2020-09-10
*//******************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/
#include "nanotube_context.hpp"
#include "nanotube_thread.hpp"

#include <cassert>
#include <cerrno>
#include <cstdio>
#include <cstring>
#include <iostream>

extern "C" {
#include <pthread.h>
#include <unistd.h>
}

static const long ns_per_s = 1000*1000*1000;
static const long test_timeout = ns_per_s/2;

pthread_mutex_t mutex_1;
pthread_cond_t cond_1;
int state = -1;

nanotube_thread *thread_ptr_1;
nanotube_thread *thread_ptr_2;

void fatal_error(const std::string &message, int rc)
{
  fprintf(stderr, "%s: %s (rc=%d)\n", message.c_str(),
          strerror(rc), rc);
  exit(1);
}

void set_state(int s)
{
  int rc;

  rc = pthread_mutex_lock(&mutex_1);
  if (rc != 0)
    fatal_error("pthread_mutex_lock failed", rc);

  state = s;
  rc = pthread_cond_broadcast(&cond_1);
  if (rc != 0)
    fatal_error("pthread_cond_broadcast", rc);

  rc = pthread_mutex_unlock(&mutex_1);
  if (rc != 0)
    fatal_error("pthread_mutex_lock failed", rc);
}

void wait_state(int s)
{
  int rc;

  rc = pthread_mutex_lock(&mutex_1);
  if (rc != 0)
    fatal_error("pthread_mutex_lock failed", rc);

  while (state != s) {
    rc = pthread_cond_wait(&cond_1, &mutex_1);
    if (rc != 0)
      fatal_error("pthread_cond_wait", rc);
  }

  rc = pthread_mutex_unlock(&mutex_1);
  if (rc != 0)
    fatal_error("pthread_mutex_lock failed", rc);
}

void func_1(nanotube_context_t *context, void *arg)
{
  assert(strcmp((char*)arg, "data_1") == 0);
  thread_ptr_1->check_current();

  std::cout << "func_1: Started.\n";

  assert(state == 0);

  // Indicate that this thread has started.
  set_state(1);

  // Wait for the second thread to wake this thread.
  nanotube_thread_wait();

  std::cout << "func_1: Woken (1).\n";

  // Make sure the second thread has woken this thread.
  assert(state == 2);

  // Indicate that this thread has woken.
  set_state(3);

  // Wait for the second thread to set a pending wake.
  wait_state(4);

  std::cout << "func_1: Sleeping again.\n";

  // This should return immediately since the wait is pending.
  nanotube_thread_wait();

  std::cout << "func_1: Woken (2).\n";

  // Indicate that the thread is about to go to sleep again.
  set_state(5);

  // Go to sleep again.
  nanotube_thread_wait();

  // Make sure the second thread has woken this thread.
  assert(state == 6);

  std::cout << "func_1: Woken (3).\n";

  // Indicate that the test is complete.
  set_state(10);

  while (true) {
    nanotube_thread_wait();
  }
}

void func_2(nanotube_context_t *context, void *arg)
{
  int rc;

  assert(strcmp((char*)arg, "data_2") == 0);
  thread_ptr_2->check_current();
  assert(state != -1);

  // Wait for the first thread to start.
  wait_state(1);

  std::cout << "func_2: Started.\n";

  // Give the first thread time to go to sleep.
  rc = usleep(1000);
  if (rc != 0)
    fatal_error("usleep", errno);

  std::cout << "func_2: Wake up! (1)\n";

  // Make sure the first thread has not woken up.
  assert(state == 1);

  // Indicate that the first thread can wake up.
  set_state(2);

  // Wake the first thread.
  thread_ptr_1->wake();

  // Wait for the first thread to wake up.
  wait_state(3);

  std::cout << "func_2: Wake up! (2)\n";

  // Wake the first thread while it is not sleeping.
  thread_ptr_1->wake();

  // Allow the first thread to go to sleep.
  set_state(4);

  // Wait for the first thread to go to sleep again.
  wait_state(5);

  std::cout << "func_2: Delaying...\n";

  // Give the first thread time to go to sleep.
  rc = usleep(1000);
  if (rc != 0)
    fatal_error("usleep", errno);

  std::cout << "func_2: Wake up! (3)\n";

  // Indicate that the first thread is allowed to wake up.
  set_state(6);

  // Wake the first thread again.
  thread_ptr_1->wake();

  while (true) {
    nanotube_thread_wait();
  }
}

int main()
{
  int rc;

  rc = pthread_mutex_init(&mutex_1, nullptr);
  if (rc != 0)
    fatal_error("pthread_mutex_init failed", rc);

  rc = pthread_cond_init(&cond_1, nullptr);
  if (rc != 0)
    fatal_error("pthread_cond_init failed", rc);

  nanotube_context main_context;
  nanotube_thread main_thread(&main_context);

  char data_1[] = "data_1";
  nanotube_context context_1;
  nanotube_thread thread_1(&context_1, "thread_1", &func_1, data_1,
                           sizeof(data_1));
  thread_ptr_1 = &thread_1;

  char data_2[] = "data_2";
  nanotube_context context_2;
  nanotube_thread thread_2(&context_2, "thread_2", &func_2, data_2,
                           sizeof(data_2));
  thread_ptr_2 = &thread_2;

  // Make sure the threads do not start before the start request.
  rc = usleep(1000);
  if (rc != 0)
    fatal_error("usleep", errno);

  // Allow the threads to start.
  state = 0;

  // Start the threads.
  thread_1.start();
  thread_2.start();

  struct timespec ts;
  rc = clock_gettime(CLOCK_REALTIME, &ts);
  if (ts.tv_nsec < ns_per_s - test_timeout) {
    ts.tv_nsec += test_timeout;
  } else {
    ts.tv_sec += 1;
    ts.tv_nsec -= (ns_per_s - test_timeout);
  }

  rc = pthread_mutex_lock(&mutex_1);
  if (rc != 0)
    fatal_error("pthread_mutex_lock failed", rc);

  while (state != 10) {
    rc = pthread_cond_timedwait(&cond_1, &mutex_1, &ts);

    if (rc == ETIMEDOUT) {
      fprintf(stderr, "Test timed out.\n");
      exit(1);
    }

    if (rc != 0)
      fatal_error("pthread_cond_timedwait", rc);
  }

  rc = pthread_mutex_unlock(&mutex_1);
  if (rc != 0)
    fatal_error("pthread_mutex_lock failed", rc);

  thread_1.stop();
  thread_2.stop();

  printf("Test complete!\n");
  
  return 0;
}
