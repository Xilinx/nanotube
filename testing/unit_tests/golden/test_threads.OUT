func_1: Started.
func_2: Started.
func_2: Wake up! (1)
func_1: Woken (1).
func_2: Wake up! (2)
func_1: Sleeping again.
func_1: Woken (2).
func_2: Delaying...
func_2: Wake up! (3)
func_1: Woken (3).
Test complete!
